File: cli/python/rendo_cli/scaffold.py
from __future__ import annotations

from datetime import datetime, timezone
import hashlib
import os
from pathlib import Path
import shutil

from .contracts import validate_project_manifest, validate_template_manifest
from .fs import copy_tree_with_replacements, ensure_missing_or_empty_dir, read_json, slugify, write_json


def resolve_selected_surfaces(manifest: dict, requested_surfaces: list[str] | None) -> list[str]:
    capabilities = manifest.get("surfaceCapabilities", [])
    if not capabilities:
        return []

    selected = requested_surfaces or manifest.get("defaultSurfaces") or ["web"]
    invalid = [surface for surface in selected if surface not in capabilities]
    if invalid:
        raise RuntimeError(f"unsupported surfaces for {manifest['id']}: {', '.join(invalid)}")
    if "web" not in selected:
        raise RuntimeError(f"current implementation requires the web surface; requested surfaces: {', '.join(selected)}")
    return list(dict.fromkeys(selected))


def prune_unselected_surfaces(target: Path, manifest: dict, selected_surfaces: list[str]) -> None:
    capabilities = manifest.get("surfaceCapabilities", [])
    surface_paths = manifest.get("surfacePaths", {})
    for surface in capabilities:
        if surface in selected_surfaces:
            continue
        for relative_path in surface_paths.get(surface, []):
            path = target / relative_path
            if path.is_dir():
                shutil.rmtree(path, ignore_errors=True)
            elif path.exists():
                path.unlink()


def build_workspace_id(template_id: str, project_name: str, created_at: str) -> str:
    slug = slugify(project_name) or "workspace"
    digest = hashlib.sha256(f"{template_id}:{slug}:{created_at}".encode("utf-8")).hexdigest()[:12]
    return f"ws-{slug}-{digest}"


def persist_workspace_metadata(
    source: dict,
    target: Path,
    selected_runtime: str,
    selected_surfaces: list[str],
    creation_mode: str,
) -> None:
    manifest = source["manifest"]
    project = validate_project_manifest(read_json(target / "rendo.project.json"))
    project["surfaces"] = selected_surfaces
    project["template"]["templateRole"] = "derived"
    project["workspaceId"] = build_workspace_id(manifest["id"], project["projectName"], project["template"]["createdAt"])
    project["origin"] = {
        "createdBy": creation_mode,
        "registry": source["registry"],
        "source": source["source"],
        "templateId": manifest["id"],
        "templateKind": manifest["templateKind"],
        "templateRole": manifest["templateRole"],
        "templateVersion": manifest["version"],
        "runtimeMode": selected_runtime,
    }
    template = validate_template_manifest(read_json(target / "rendo.template.json"))
    template["templateRole"] = "derived"
    write_json(target / ".rendo" / "rendo.project.json", project)
    write_json(target / ".rendo" / "rendo.template.json", template)
    (target / "rendo.project.json").unlink(missing_ok=True)
    (target / "rendo.template.json").unlink(missing_ok=True)


def scaffold_template(
    source: dict,
    target_dir: str,
    runtime_mode: str | None = None,
    requested_surfaces: list[str] | None = None,
    creation_mode: str = "rendo.create",
) -> dict:
    target = Path(target_dir).resolve()
    ensure_missing_or_empty_dir(target)

    manifest = source["manifest"]
    selected_runtime = runtime_mode or manifest["runtimeModes"][0]
    selected_surfaces = resolve_selected_surfaces(manifest, requested_surfaces)
    project_name = target.name
    created_at = os.environ.get("RENDO_CREATED_AT_OVERRIDE", datetime.now(timezone.utc).isoformat())
    replacements = {
        "__RENDO_PROJECT_NAME__": project_name,
        "__RENDO_PROJECT_SLUG__": slugify(project_name),
        "__RENDO_TEMPLATE_ID__": manifest["id"],
        "__RENDO_TEMPLATE_TITLE__": manifest["title"],
        "__RENDO_TEMPLATE_KIND__": manifest["templateKind"],
        "__RENDO_TEMPLATE_ROLE__": manifest["templateRole"],
        "__RENDO_TEMPLATE_VERSION__": manifest["version"],
        "__RENDO_RUNTIME_MODE__": selected_runtime,
        "__RENDO_CREATED_AT__": created_at,
        "__RENDO_SELECTED_SURFACES__": ", ".join(selected_surfaces) if selected_surfaces else "none",
    }
    copied = copy_tree_with_replacements(source["templateDir"], target, replacements)
    prune_unselected_surfaces(target, manifest, selected_surfaces)
    persist_workspace_metadata(source, target, selected_runtime, selected_surfaces, creation_mode)
    normalized_copied = [
        relative_path
        for relative_path in copied
        if relative_path not in {"rendo.project.json", "rendo.template.json"}
    ] + [
        ".rendo/rendo.project.json",
        ".rendo/rendo.template.json",
    ]
    return {
        "targetDir": str(target),
        "templateId": manifest["id"],
        "copiedFiles": normalized_copied,
        "selectedSurfaces": selected_surfaces,
        "nextSteps": [f"cd {target}", "npm install", "npm run health", "npm run check"],
    }

File: rendo-cli/cli/node/src/scaffold.ts
import { createHash } from "node:crypto";
import path from "node:path";
import { copyTreeWithReplacements, ensureMissingOrEmptyDir, slugify, writeJsonFile } from "./fs.js";
import { projectManifestSchema, templateManifestSchema, type TemplateManifest } from "./contracts.js";
import { readJsonFile } from "./fs.js";
import { promises as fs } from "node:fs";
import type { PreparedTemplateSource } from "./registry-client.js";

export type ScaffoldResult = {
  targetDir: string;
  templateId: string;
  copiedFiles: string[];
  selectedSurfaces: string[];
  nextSteps: string[];
};

export type WorkspaceCreationMode = "rendo.init" | "rendo.create" | "rendo.pull";
export type TemplateWorkspaceOptions = {
  asTemplate?: boolean;
  templateId?: string;
  templateName?: string;
  templateTitle?: string;
};

function humanizeTemplateId(templateId: string) {
  return templateId
    .replaceAll("/", "-")
    .replaceAll("_", "-")
    .split("-")
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ") || "Derived Template";
}

function resolveSelectedSurfaces(manifest: TemplateManifest, requestedSurfaces?: string[]) {
  if (manifest.surfaceCapabilities.length === 0) {
    return [];
  }

  const selected = requestedSurfaces?.length
    ? requestedSurfaces
    : manifest.defaultSurfaces.length
      ? manifest.defaultSurfaces
      : ["web"];

  const invalid = selected.filter((surface) => !manifest.surfaceCapabilities.includes(surface as any));
  if (invalid.length > 0) {
    throw new Error(`unsupported surfaces for ${manifest.id}: ${invalid.join(", ")}`);
  }

  if (!selected.includes("web")) {
    throw new Error(`current implementation requires the web surface; requested surfaces: ${selected.join(", ")}`);
  }

  return [...new Set(selected)];
}

async function pruneUnselectedSurfaces(
  targetDir: string,
  manifest: TemplateManifest,
  selectedSurfaces: string[],
): Promise<void> {
  if (manifest.surfaceCapabilities.length === 0) {
    return;
  }

  for (const surface of manifest.surfaceCapabilities) {
    if (selectedSurfaces.includes(surface)) {
      continue;
    }
    const paths = manifest.surfacePaths[surface] ?? [];
    for (const relativePath of paths) {
      await fs.rm(path.join(targetDir, relativePath), { recursive: true, force: true, maxRetries: 10, retryDelay: 50 });
    }
  }
}

function buildWorkspaceId(templateId: string, projectName: string, createdAt: string) {
  const slug = slugify(projectName) || "workspace";
  const digest = createHash("sha256")
    .update(`${templateId}:${slug}:${createdAt}`)
    .digest("hex")
    .slice(0, 12);
  return `ws-${slug}-${digest}`;
}

async function persistWorkspaceMetadata(
  source: PreparedTemplateSource,
  targetDir: string,
  selectedRuntimeMode: string,
  selectedSurfaces: string[],
  creationMode: WorkspaceCreationMode,
  localTemplateIdentity: { id: string; name: string; title: string },
) {
  const projectManifestPath = path.join(targetDir, "rendo.project.json");
  const templateManifestPath = path.join(targetDir, "rendo.template.json");
  const workspaceProjectManifestPath = path.join(targetDir, ".rendo", "rendo.project.json");
  const workspaceTemplateManifestPath = path.join(targetDir, ".rendo", "rendo.template.json");
  const manifest = source.manifest;
  const project = projectManifestSchema.parse(await readJsonFile(projectManifestPath));
  project.surfaces = selectedSurfaces as any;
  project.template.id = localTemplateIdentity.id;
  project.template.templateRole = "derived";
  project.workspaceId = buildWorkspaceId(localTemplateIdentity.id, project.projectName, project.template.createdAt);
  project.origin = {
    createdBy: creationMode,
    registry: source.registry,
    source: source.source,
    templateId: manifest.id,
    templateKind: manifest.templateKind,
    templateRole: manifest.templateRole,
    templateVersion: manifest.version,
    runtimeMode: selectedRuntimeMode as any,
  };

  const template = templateManifestSchema.parse(await readJsonFile(templateManifestPath));
  template.id = localTemplateIdentity.id;
  template.name = localTemplateIdentity.name;
  template.title = localTemplateIdentity.title;
  template.templateRole = "derived";

  await writeJsonFile(workspaceProjectManifestPath, project);
  await writeJsonFile(workspaceTemplateManifestPath, template);
  await fs.rm(projectManifestPath, { force: true });
  await fs.rm(templateManifestPath, { force: true });
}

export async function scaffoldTemplate(
  source: PreparedTemplateSource,
  requestedTargetDir: string,
  runtimeMode?: string,
  requestedSurfaces?: string[],
  creationMode: WorkspaceCreationMode = "rendo.create",
  templateWorkspaceOptions: TemplateWorkspaceOptions = {},
): Promise<ScaffoldResult> {
  const targetDir = path.resolve(requestedTargetDir);
  await ensureMissingOrEmptyDir(targetDir);
  const manifest = source.manifest;
  const projectName = path.basename(targetDir);
  const projectSlug = slugify(projectName);
  const selectedRuntimeMode = runtimeMode ?? manifest.runtimeModes[0];
  const createdAt = process.env.RENDO_CREATED_AT_OVERRIDE ?? new Date().toISOString();
  const selectedSurfaces = resolveSelectedSurfaces(manifest, requestedSurfaces);
  if (templateWorkspaceOptions.asTemplate && !templateWorkspaceOptions.templateId) {
    throw new Error("template workspaces require --template-id");
  }
  const localTemplateIdentity = {
    id: templateWorkspaceOptions.templateId ?? manifest.id,
    name: templateWorkspaceOptions.templateName ?? humanizeTemplateId(templateWorkspaceOptions.templateId ?? manifest.id),
    title: templateWorkspaceOptions.templateTitle ?? templateWorkspaceOptions.templateName ?? humanizeTemplateId(templateWorkspaceOptions.templateId ?? manifest.id),
  };
  const copiedFiles = await copyTreeWithReplacements(
    source.templateDir,
    targetDir,
    {
      "__RENDO_PROJECT_NAME__": projectName,
      "__RENDO_PROJECT_SLUG__": projectSlug,
      "__RENDO_TEMPLATE_ID__": localTemplateIdentity.id,
      "__RENDO_TEMPLATE_TITLE__": localTemplateIdentity.title,
      "__RENDO_TEMPLATE_KIND__": manifest.templateKind,
      "__RENDO_TEMPLATE_ROLE__": manifest.templateRole,
      "__RENDO_TEMPLATE_VERSION__": manifest.version,
      "__RENDO_RUNTIME_MODE__": selectedRuntimeMode,
      "__RENDO_CREATED_AT__": createdAt,
      "__RENDO_SELECTED_SURFACES__": selectedSurfaces.join(", ") || "none",
    },
  );
  await pruneUnselectedSurfaces(targetDir, manifest, selectedSurfaces);
  await persistWorkspaceMetadata(source, targetDir, selectedRuntimeMode, selectedSurfaces, creationMode, localTemplateIdentity);
  const normalizedCopiedFiles = copiedFiles
    .filter((relativePath) => relativePath !== "rendo.project.json" && relativePath !== "rendo.template.json")
    .concat([".rendo/rendo.project.json", ".rendo/rendo.template.json"]);

  const nextSteps = [
    `cd ${targetDir}`,
    "npm install",
    "npm run health",
    "npm run check",
  ];

  return {
    targetDir,
    templateId: localTemplateIdentity.id,
    copiedFiles: normalizedCopiedFiles,
    selectedSurfaces,
    nextSteps,
  };
}
